casper/src/main/scala/coop/rchain/casper/blocks/proposer/Proposer.scala

```diff
@@ -128,10 +131,76 @@ object Proposer {
         maxSeqNum  = latestMsgs.find(_.sender == sender).map(_.senderSeq)
       } yield maxSeqNum.getOrElse(-1)
 
-    def createBlock(validatorIdentity: ValidatorIdentity) =
+    def createBlock(validatorIdentity: ValidatorIdentity): F[BlockCreatorResult] =
       for {
-        parentsPreState <- MultiParentCasper.getPreStateForNewBlock
-        result          <- BlockCreator.create(parentsPreState, validatorIdentity, shardId, dummyDeployOpt)
+        // merge pre state
+        preState <- MultiParentCasper.getPreStateForNewBlock
+        // misc
+        preStateHash      = preState.preStateHash
+        creatorsPk        = validatorIdentity.publicKey
+        creatorsId        = ByteString.copyFrom(creatorsPk.bytes)
+        creatorsLatestOpt = preState.justifications.find(_.sender == creatorsId)
+        nextSeqNum        = creatorsLatestOpt.map(_.seqNum + 1).getOrElse(0L)
+        nextBlockNum      = preState.justifications.map(_.blockNum).max + 1
+        parentHashes      = preState.justifications.map(_.blockHash)
+        finalBonds        = preState.fringeBondsMap
+        offenders         = preState.justifications.filter(_.validationFailed).map(_.sender)
+        // slashing
+        preStateBonds <- RuntimeManager[F].computeBonds(preStateHash.toByteString)
```

Comment:
We cannot use bonds from PoS here, previous code is wrong in doing this. PoS bonds should only be used in epoch change to add new validators to _final_ bonds and to remove rejected from _final_ bonds.

Bonds on final state + bonds from consensus (parents) together must be used to validate and generate new blocks. Bonds from final state defines all active validators and bonds from consensus defines who is rejected in non-finalized blocks until the new bonds is finalized.

For now just use bonds from pre-state.

Reply:
I'm not sure wdy mean - `For now just use bonds from pre-state.` I'm doing exactly that, don't I?

Reply:
It's fine for now until finalizer have support for invalid blocks.

Situation that I have in mind is just before epoch change is finalized and fringe bonds map contains old validators and pre-state bonds contains new validators. When old validator is slashed it may not be visible to pre-state bonds, I'm not sure how this will be handled in PoS. Maybe I'm missing something now.